doc/platform-optimization/README.md

```diff
@@ -0,0 +1,188 @@
+there are "tgi" and "tei" containers that will need a lot of CPUs.
+
+We will allocate 16 CPUs (8 full CPU cores with two hyperthreads on
+each core) for each tgi container, and 32 CPUs (that is 16
+hyperthreaded cores) for each tei container. This is done with the
+following balloons policy configuration that is explained below.
+
+```yaml
+apiVersion: config.nri/v1alpha1
+kind: BalloonsPolicy
+metadata:
+  name: default
+spec:
+  allocatorTopologyBalancing: true
+  balloonTypes:
+  - name: tgi
+    allocatorPriority: high
+    minCPUs: 16
+    minBalloons: 1
+    preferNewBalloons: true
+    hideHyperthreads: true
+    matchExpressions:
+    - key: name
+      operator: Equals
+      values: ["tgi"]
+  - name: tei
+    allocatorPriority: high
+    minCPUs: 32
+    minBalloons: 1
+    preferNewBalloons: true
+    hideHyperthreads: true
+    matchExpressions:
+    - key: name
+      operator: Equals
+      values: ["tei"]
+  - allocatorPriority: normal
+    minCPUs: 1
+    hideHyperthreads: false
+    name: default
+    namespaces:
+    - "*"
+    shareIdleCPUsInSame: numa
+  instrumentation:
+    httpEndpoint: :8891
+    prometheusExport: true
+    reportPeriod: 60s
+    samplingRatePerMillion: 0
+  log:
+    source: true
+    debug: ["policy"]
+  pinCPU: true
+  pinMemory: false
+  reservedPoolNamespaces:
+  - kube-system
+  reservedResources:
+    cpu: "2"
+```
+
+The balloons policy creates "balloons" of CPUs that only containers
+assigned into a balloon are allowed to use.
+
+The most important options in the above configuration example are:
+
+- `allocatorTopologyBalancing: true`. This option ensures that
+  balloons (sets of allowed CPUs) are balanced between CPU sockets in
+  the system, and local NUMA nodes if the system is running a sub-NUMA
+  clustering (SNC) mode. Without this options balloons would be
+  tightly packed on a single socket allowing the other CPU socket
+  sleep and save power. Here we have optimized for performance.
+- The list of `balloonTypes` includes two application-specific balloon
+  types: one for tgi and one for tei containers.
+- `matchExpressions` of a balloon type enable matching containers that
+  should be run in balloons of this type. We select tei and tgi
+  containers into their special balloon types based on container
+  name. Matching could be done based on labels and pod name, too.
+- `preferNewBalloon: true` on both tei and tgi balloon types means
+  that when a container is assigned into this balloon type and it is
+  possible to create a new balloon of this type because there are
+  enough free CPUs in the system, then the new balloon will be created
+  for the container. As a result, both tei and tgi containers will get
+  dedicated set of CPUs, unlike other containers that will run in the
+  default balloon type. Each container is allowed to use only CPUs of
+  the balloon where they are assigned.
```

Comment:
With new, I am assuming from your description that a given CPU, physical or hyperthreaded is allocated to only one balloon.

is there a performance penalty when using a balloon for a workload that spans NUMA boundaries .. all the cache and memory reference related added latency?

Reply:
Your assumption is correct. I also added a clarification on that: a balloon can belong to at most one CPU at a time.

Spanning over NUMA boundaries is quite likely to hurt performance, especially if NUMA nodes are in different sockets. Sometimes it's better to have fewer CPUs and all data in local memory than more CPUs and a lot of memory accesses to remote NUMA.